Add MaxOut layer (#647)
FluxML/Flux.jl: docs/src/models/layers.md

Flux.Recur
```

## Hipster Layers

> These are marginally more obscure layers that you probably haven't heard of.

Reviewer: I know you mean it as a joke, but this can come out as condescending.

Author: I put it in knowing that in code reveiw someone would complain and hopefully come up with something better.

Reviewer 2: Just "Other"?

Author: Jules Merckx   came up with "esoteric layers",
which I like better than Other, it is like Misc
Anything could be in Other where as Esoteric says it is not the run of the mill layer.

Reviewer 2: Do you have a list of those layers that will classify as "esoteric"? Seems like such a notation will depend on the person and the application. This seciton will likely effectively just work as an "Other" section so might as well call it that.

Author: I want a way to be clear that these are not domain specific ones.
In contrast to RNN or CNN.

Perhaps "Other General Purpose Layers" ?

Flux.Recur
```

## Other General Purpose Layers
These are marginally more obscure than the Basic Layers.

> But incontrast to the layers described in the other sections are not readily grouped around a paparticular purpose (e.g. CNNs or RNNs).

Reviewer 3: ```suggestion
But in contrast to the layers described in the other sections are not readily grouped around a paparticular purpose (e.g. CNNs or RNNs).
```

> But incontrast to the layers described in the other sections are not readily grouped around a paparticular purpose (e.g. CNNs or RNNs).

Reviewer 4: ```suggestion
But in contrast to the layers described in the other sections are not readily grouped around a particular purpose (e.g. CNNs or RNNs).
```

```@docs

> MaxOut

Reviewer 4: ```suggestion
Maxout
```
